core/src/impl/Kokkos_Core.cpp

```diff
@@ -151,6 +155,32 @@ void combine(Kokkos::Tools::InitArguments& out,
   }
 }
 
+int get_device_count() {
+#if defined(KOKKOS_ENABLE_CUDA)
+  return Kokkos::Cuda::detect_device_count();
+#elif defined(KOKKOS_ENABLE_HIP)
+  return Kokkos::Experimental::HIP::detect_device_count();
+#elif defined(KOKKOS_ENABLE_SYCL)
+  return sycl::device::get_devices(sycl::info::device_type::gpu).size();
+#else
+  // This function is always compiled but should only be ever called when
+  // either CUDA, HIP, or SYCL are enabled.
+  return *reinterpret_cast<int*>(0x8BADF00D);  // implementation bug
```

Comment:
why this? why it says implementation bug?

Reply:
`get_device_count() -> int` is only implemented for CUDA/HIP/SYCL but the function actually gets compiled regardless.  It is called from `Impl::get_gpu(InitializationSettings const&) -> int` which gets only used in these 3 backends.  I wrote `return *reinterpret_cast<int*>(0x8BADF00D);` to cause a segfault if that changes (say OpenMPTarget or OpenACC attempt to call `get_gpu`) so that we don't silently compute garbage and it is obvious where the issue is coming from.

Reply:
If I return say `-1` one would see errors/warnings saying there is no device available and would have to trace back to that function.  I opted for failing early in an attempt to make it easier to debug.  Does that make sense?

Reply:
ok, i see. Why not just aborting instead of this reinterpret cast? 
Also, I think it would be great to have your explanation above added as comment there so that one clearly understands things

Reply:
>Speculating I would get warnings that function does not return but I can change if you request me to.
ok I am fine keeping the segfault 

currently we have:
```cpp
// This function is always compiled but should only be ever called when
// either CUDA, HIP, or SYCL are enabled.
```
I would suggest changing to:
```cpp
/* get_device_count() -> int is only implemented for CUDA/HIP/SYCL but the function actually gets 
    compiled regardless. It is called from Impl::get_gpu(InitializationSettings const&) -> int 
   which gets only used in these 3 backends. I wrote return *reinterpret_cast<int*>(0x8BADF00D); 
   to cause a segfault if that changes (say OpenMPTarget or OpenACC attempt to call get_gpu) 
   so that we don't silently compute garbage and it is obvious where the issue is coming from. */
```